.ci/load/Jenkinsfile

```diff
@@ -0,0 +1,169 @@
+        string(name: "concurrent_requests", defaultValue: "100", description: "The number of concurrent requests to test with")
+        string(name: "duration", defaultValue: "10", description: "Test duration in minutes. Max: 280")
+        // num_of_runs currently unsupported
+        // string(name: "num_of_runs", defaultValue: "1", description: "Number of test runs to execute")
+        text(name: "agent_config", "defaultValue": "", description: "Custom APM Agent configuration. (WARNING: May echo to console. Do not supply sensitive data.)")
+        text(name: "locustfile", "defaultValue": "", description: "Locust load-generator plan")
+        booleanParam(name: "local_metrics", description: "Enable local metrics collection?", defaultValue: false)
+        // End script auto-generation
+    }
+
+    stages {
+        stage('Pre-flight'){
+            steps {
+                echo 'Getting authentication information from Vault'
+                withSecretVault(secret: 'secret/apm-team/ci/bandstand', user_var_name: 'APP_TOKEN_TYPE', pass_var_name: 'APP_TOKEN'){
+                    setEnvVar('SESSION_TOKEN', sh(script: ".ci/load/scripts/start.sh", returnStdout: true).trim())
+                }
+            }    
+        }
+        stage('Load test') {
+            parallel {
+                stage('Load generation') {
+                    agent { label 'metal' }
+                    steps {
+                            withSecretVault(secret: 'secret/apm-team/ci/bandstand', user_var_name: 'APP_TOKEN_TYPE', pass_var_name: 'APP_TOKEN'){
+                                echo 'Preparing load generation..'
+                                whenTrue(Boolean.valueOf(params.locustfile)) {
+                                    echo 'Using user-supplied plan for load-generation with Locust'
+                                    sh script: "echo \"${params.locustfile}\">.ci/load/scripts/locustfile.py"
+                                }
+                                sh(script: ".ci/load/scripts/load_agent.sh")
+                            }
+                    }
+                }
+                stage('Test application') {
+                    agent { label 'metal' }
```

Comment:
are the workers guaranteed to be the same run-to-run? how many cpu cores do they have?

Reply:
This is an important discussion point. Thanks for bringing it up!

The tl;dr here is that they _should_ be the same but an absolute guarantee is challenging. For example, the ES Performance Testing team has a fleet of machines but have discovered over time that there are going to be variances even when they try hard to avoid them. SSDs in arrays fail, machines aren't tagged the same way by the provider but aren't physically identical, etc, etc.

So, where does that leave us? I think the best thing to do here is to be cautious about comparing results _between_ runs but that we continue to enhance this pipeline to support scenarios where we can run multiple invocations of a single test scenario multiple times on what we can guarantee to be the same machine(s) and maybe even some sort of comparative logic as well. (So, run scenario A and the scenario B on the same machine and output the results.)

I'm also going to file an issue in the infra repo to try to get an audit underway so we can know a bit more about what divergence we do have currently. As mentioned earlier, it should be a _lot_ but it may be some.

Reply:
Some additional data here:

We have four workers right now, all of which vary in subtle ways. I propose that we make the following changes:

1. Pin the stage that runs the application to the same worker every time. We can do this by using the `benchmark` label, which currently is only assigned to one machine. That machine has the following specs:

*  Ubuntu 18.04
* 6CPUs 
* 64 GB 

2. We keep the load-generation stage marked as `metal` which will allow it to float between the other bare-metal machines which have slightly varying specs. However, there's not much reason to believe that they vary enough to modify the behavior of the load-generation script, which doesn't consume a great deal of resources at present.

3. We decide on a plan to order some additional machines which we can put into the `benchmark` pool which will ensure better consistency going forward.

(I will link backward from a ticket which provides more info, so we don't link from a public repo into a private one.)

Let me know how this sounds @felixbarny and @v1v and if you give a 👍 I will make the necessary change.

Reply:
👍